vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -0,0 +1,785 @@
+#include "conditional_contextual_bandit.h"
+#include "reductions.h"
+#include "example.h"
+#include "global_data.h"
+#include "cache.h"
+#include "vw.h"
+#include "interactions.h"
+#include "label_dictionary.h"
+#include "cb_adf.h"
+#include "cb_algs.h"
+
+#include <numeric>
+#include <algorithm>
+#include <unordered_set>
+#include <unordered_map>
+
+using namespace LEARNER;
+using namespace VW;
+using namespace VW::config;
+
+
+struct ccb
+{
+  example* shared;
+  std::vector<example*> actions, decisions;
+  uint32_t chosen_action_index;
+  std::unordered_map<uint32_t, uint32_t> origin_index;
```

Comment:
Given the key to this map will be a number known to be in the range `[0 - data.actions.size() [`  we can use a fully associative type instead. This should be a `std::vector<uint32_t>`.